ibm/data_source_ibm_kms_key_rings.go

```diff
@@ -0,0 +1,162 @@
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Key protect or hpcs instance GUID",
+			},
+			"endpoint_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateAllowedStringValue([]string{"public", "private"}),
+				Description:  "public or private",
+				Default:      "public",
+			},
+			"key_rings": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"creation_date": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"created_by": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceIBMKMSKeyRingsRead(d *schema.ResourceData, meta interface{}) error {
+	api, err := meta.(ClientSession).keyManagementAPI()
+	if err != nil {
+		return err
+	}
+
+	rContollerClient, err := meta.(ClientSession).ResourceControllerAPIV2()
+	if err != nil {
+		return err
+	}
+
+	instanceID := d.Get("instance_id").(string)
+	endpointType := d.Get("endpoint_type").(string)
+
+	rContollerApi := rContollerClient.ResourceServiceInstanceV2()
+
+	instanceData, err := rContollerApi.GetInstance(instanceID)
+	if err != nil {
+		return err
+	}
+	instanceCRN := instanceData.Crn.String()
+
+	var hpcsEndpointURL string
+	crnData := strings.Split(instanceCRN, ":")
+
+	if crnData[4] == "hs-crypto" {
+
+		hpcsEndpointApi, err := meta.(ClientSession).HpcsEndpointAPI()
+		if err != nil {
+			return err
+		}
+		resp, err := hpcsEndpointApi.Endpoint().GetAPIEndpoint(instanceID)
+		if err != nil {
+			return err
+		}
+
+		if endpointType == "public" {
+			hpcsEndpointURL = "https://" + resp.Kms.Public + "/api/v2/keys"
+		} else {
+			hpcsEndpointURL = "https://" + resp.Kms.Private + "/api/v2/keys"
+		}
+
+		u, err := url.Parse(hpcsEndpointURL)
+		if err != nil {
+			return fmt.Errorf("Error Parsing hpcs EndpointURL")
+		}
+		api.URL = u
+	} else if crnData[4] == "kms" {
+		if endpointType == "private" {
+			if !strings.HasPrefix(api.Config.BaseURL, "private") {
+				api.Config.BaseURL = "private." + api.Config.BaseURL
+			}
+		}
+	} else {
+		return fmt.Errorf("Invalid or unsupported service Instance")
+	}
+
+	api.Config.InstanceID = instanceID
+	keys, err := api.GetKeyRings(context.Background())
+	if err != nil {
+		return fmt.Errorf(
+			"Get Keys failed with error: %s", err)
+	}
+	retreivedKeyRings := keys.KeyRings
+	if len(retreivedKeyRings) == 0 {
+		return fmt.Errorf("No keys in instance  %s", instanceID)
+	}
+	var keyName string
+	var matchKeys []kp.KeyRing
+	for _, keyRingData := range retreivedKeyRings {
+		matchKeys = append(matchKeys, keyRingData)
+	}
+
+	if len(matchKeys) == 0 {
+		return fmt.Errorf("No keys with name %s in instance  %s", keyName, instanceID)
+	}
+
+	keyMap := make([]map[string]interface{}, 0, len(matchKeys))
```

Comment:
```
var matchKeys []kp.KeyRing
	for _, keyRingData := range retreivedKeyRings {
		matchKeys = append(matchKeys, keyRingData)
	}

	if len(matchKeys) == 0 {
		return fmt.Errorf("No keys with name %s in instance  %s", keyName, instanceID)
	}
```
construction of `matchKeys` is not required.. instead it could have directly been
`for _, key := range matchKeys`  ===>>  `for _, key := range retreivedKeyRings {`